src/shell/test.rs

```diff
@@ -1478,19 +1481,54 @@ async fn cross_platform_shebang() {
 }
 
 #[tokio::test]
-async fn provided_token_cancel() {
-  let token = CancellationToken::new();
-  token.cancel();
+async fn provided_signal_cancel() {
   let start_time = Instant::now();
-  TestBuilder::new()
-    .command("sleep 5")
-    .token(token)
-    .assert_exit_code(130)
-    .run()
-    .await;
+  {
+    let kill_signal = KillSignal::default();
+    kill_signal.send(SignalKind::SIGINT);
+    TestBuilder::new()
+      .command("sleep 5")
+      .kill_signal(kill_signal)
+      .assert_exit_code(130)
+      .run()
+      .await;
+  }
+  {
+    let kill_signal = KillSignal::default();
+    kill_signal.send(SignalKind::SIGKILL);
+    TestBuilder::new()
+      .command("sleep 5")
+      .kill_signal(kill_signal)
+      .assert_exit_code(137)
+      .run()
+      .await;
+  }
   assert!(start_time.elapsed().as_secs() < 1);
 }
 
+#[tokio::test]
+async fn listens_for_signals_exits_gracefully() {
+  if cfg!(windows) {
+    return; // signals are terrible on windows
+  }
+
+  let kill_signal = KillSignal::default();
+  deno_unsync::spawn({
+    let kill_signal = kill_signal.clone();
+    async move {
+      tokio::time::sleep(Duration::from_millis(100)).await;
+      kill_signal.send(SignalKind::SIGINT);
+    }
+  });
+  TestBuilder::new()
+      .command(r#"deno eval 'Deno.addSignalListener("SIGINT", () => { console.log("interrupted!"); setTimeout(() => Deno.exit(0), 25); }); setInterval(() => {}, 1000)'"#)
+      .kill_signal(kill_signal)
+      .assert_exit_code(0) // signal was handled and a `Deno.exit(0)` was done
```

Comment:
I improved the handling to return `0` in this case, which makes more sense because the signal was handled.